vcd/resource_vcd_api_token.go

```
@@ -0,0 +1,160 @@
+			},
+		},
+	}
+}
+
+func resourceVcdApiTokenCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	// System Admin can't create API tokens outside SysOrg,
+	// just as Org admins can't create API tokens in other Orgs
+	org := vcdClient.SysOrg
+	if org == "" {
+		org = vcdClient.Org
+	}
+
+	tokenName := d.Get("name").(string)
+	token, err := vcdClient.CreateToken(org, tokenName)
+	if err != nil {
+		return diag.Errorf("[API token create] error creating API token: %s", err)
+	}
+	d.SetId(token.Token.ID)
+
+	apiToken, err := token.GetInitialApiToken()
+	if err != nil {
+		return diag.Errorf("[API token create] error getting refresh token from API token: %s", err)
+	}
+
+	filename := d.Get("file_name").(string)
+	allowTokenFile := d.Get("allow_token_file").(bool)
+
+	var diagnostics diag.Diagnostics
+	if !allowTokenFile {
+		diagnostics = append(diagnostics, diag.Diagnostic{
+			Severity: diag.Warning,
+			Summary:  "The file " + filename + " should be considered sensitive information.",
+			Detail: "The file " + filename + " contains an API token which is " +
+				"usable by anyone to run operations to the current VCD. As such, it should be considered SENSITIVE INFORMATION. " +
+				"If you would like to remove this warning, add\n\n" + "	allow_token_file = true\n\nto the provider settings.",
+		})
+	}
+
+	if filename == "" {
+		return diag.Errorf("[API token create] file_name must be set on creation")
+	}
+
+	err = govcd.SaveApiTokenToFile(filename, vcdClient.Client.UserAgent, apiToken)
+	if err != nil {
+		return diag.Errorf("[API token create] error saving API token to file: %s", err)
+	}
+
+	return append(diagnostics, resourceVcdApiTokenRead(ctx, d, meta)...)
+}
+
+// There are no fields that can be updated after creation
+func resourceVcdApiTokenUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	return nil
+}
```

Hm, what if I decide to `allow_token_file` in an "Update" scenario?

If you don't want to support `Update` operation at all it is safest to remove that function and its usage here https://github.com/vmware/terraform-provider-vcd/pull/1070/files#diff-903e55ac1ebb7d457f38d2e93e4a27849755563c93f8b415a120b9734793868fR17

That way Terraform will self-validate that all fields have `ForceNew`, because if they don't - it means that `Update` does something

---

Thanks for pointing this out, I just wanted to not have one more API call. I changed it so it would call Read, but maybe there's a more optimal way. Didn't want to remove it completely as ForceNew on the `allow_token_file` attribute doesn't seem right.

---

But if I change `allow_token_file` on Update, will it do anything? AFAIU, you can only get the token during "create" operation, not on any subsequent "read" operations (just like in UI - shown once). 

To me it looks that it is correct behavior to have `ForceNew: true` for `allow_token_file` field as well unless I misunderstand how it works.

---

Changed as we discussed, removed the `Update` function. Also, made `allow_token_file = false` return an error instead of a warning on Service Account activation (Create or Update that sets `active = true`), so it can work on imports properly.